sw/host/rom_ext_image_tools/signer/src/main.rs

```diff
@@ -71,128 +82,111 @@ fn parse_hex_str(hex_str: &str) -> Result<Vec<u8>> {
+        .copy_from_slice(&modulus);
+    image.manifest.extension0_offset = 0;
+    image.manifest.extension0_checksum = 0;
+    image.manifest.extension1_offset = 0;
+    image.manifest.extension1_checksum = 0;
+    image.manifest.extension2_offset = 0;
+    image.manifest.extension2_checksum = 0;
+    image.manifest.extension3_offset = 0;
+    image.manifest.extension3_checksum = 0;
 
     Ok(())
 }
 
 /// Calculates the signature for the signed portion of an image.
 fn calculate_image_signature(
     image: &Image,
-    key: impl AsRef<Path>,
+    private_key: &PrivateKey,
 ) -> Result<ImageSignature> {
-    let key = fs::read(&key).with_context(|| {
-        format!("Failed to read key from `{}`.", key.as_ref().display())
-    })?;
-    let key =
-        PrivateKey::parse_from_der(&key).context("Failed to parse the key.")?;
-    let sig = ImageSignature::sign(&key, image.signed_bytes())
-        .context("Failed to sign the image.")?;
-    Ok(sig)
+    ImageSignature::sign(&private_key, &image.signed_bytes())
+        .context("Failed to calculate image signature.")
 }
 
 /// Updates the signature of an image.
-fn update_image_signature(
-    image: &mut Image,
-    signature: ImageSignature,
-) -> Result<()> {
-    image.set_manifest_field(
-        &manifest::ROM_EXT_IMAGE_SIGNATURE,
-        signature.bytes().iter().rev().cloned(),
-    )?;
-    Ok(())
+fn update_image_signature(image: &mut Image, sig: ImageSignature) {
+    image
+        .manifest
+        .signature
+        .data
+        .as_bytes_mut()
+        .copy_from_slice(
+            sig.bytes()
+                .iter()
+                .rev()
+                .cloned()
+                .collect::<Vec<_>>()
+                .as_slice(),
+        );
 }
 
 fn main() -> Result<()> {
+    // TODO(#6915): Convert this to a unit test after we start running rust tests during our
+    // builds.
+    manifest::check_manifest_layout();
     let args = Args::new(env::args_os())?;
-    let mut image = Image::from(fs::read(&args.input).with_context(|| {
-        format!("Failed to read the image from {}", args.input.display())
-    })?);
-
-    // TODO for a future refactor into opentitantool: These functions probably should belong to a
-    // Signer struct and should move into the image crate.
-    // TODO: This must use the public key.
-    update_image_manifest(&mut image, &args.priv_key)?;
-    let sig = calculate_image_signature(&image, &args.priv_key)?;
-    update_image_signature(&mut image, sig)?;
-    fs::write(&args.output, image).with_context(|| {
+    // Read image payload.
+    let mut input_file = OpenOptions::new()
+        .read(true)
+        .open(&args.input)
+        .with_context(|| format!("Failed to open {}", args.output.display()))?;
+    let file_size = input_file.metadata()?.len();
+    let mut payload =
+        Vec::with_capacity(file_size as usize - size_of::<Manifest>());
+    input_file.seek(SeekFrom::Start(u64::try_from(size_of::<Manifest>())?))?;
```

Comment:
I'd just do a cast here instead of fussing with try_from?

(Also, frankly, I wouldn't fuss with seek and just read the whole file, slicing off the leading manifest with a slice op.)